lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -170,109 +241,265 @@ struct IMConstPropPass : public IMConstPropBase<IMConstPropPass> {
     return executableBlocks.count(block);
   }
 
-  bool isOverdefined(Value value) const {
-    auto it = latticeValues.find(value);
+  bool isOverdefined(FieldRef fieldRef) const {
+    auto it = translateAndFind(fieldRef);
     return it != latticeValues.end() && it->second.isOverdefined();
   }
 
-  /// Mark the given value as overdefined. This means that we cannot refine a
-  /// specific constant for this value.
-  void markOverdefined(Value value) {
-    auto &entry = latticeValues[value];
+  // Value is regarded as overdefined if one of lattice values associated with
+  // the value is overdefined.
+  bool isOverdefined(Value value) const {
+    auto [root, index, size] = getRootValueWithCorrespondingLeafIDRange(value);
+    auto types = getLeafFieldIDsAndGroundTypes(root.getType());
+    for (unsigned i = 0; i < size; ++i)
+      if (isOverdefined({root, types[index + i].first}))
+        return true;
+    return false;
+  }
+
+  /// Mark the given fieldRef as overdefined. This means that we cannot refine a
+  /// specific constant for this fieldRef.
+  void markOverdefined(FieldRef fieldRef) {
+    fieldRef = getFieldRefFromFieldRef(fieldRef);
+    auto &entry = latticeValues[fieldRef];
     if (!entry.isOverdefined()) {
       entry.markOverdefined();
-      changedLatticeValueWorklist.push(value);
+      changedLatticeValueWorklist.push(fieldRef);
     }
   }
 
+  /// Mark the given value as overdefined. This means that we cannot
+  /// refine a specific constant for this value. The value might be associated
+  /// with some values if the value has a aggregate type so iterate over the
+  /// corresponding range.
+  void markOverdefined(Value value) {
+    auto [root, index, size] = getRootValueWithCorrespondingLeafIDRange(value);
+    auto types = getLeafFieldIDsAndGroundTypes(root.getType());
+    for (unsigned int i = 0; i < size; ++i)
+      markOverdefined({root, types[index + i].first});
+  }
+
   /// Merge information from the 'from' lattice value into value.  If it
   /// changes, then users of the value are added to the worklist for
   /// revisitation.
-  void mergeLatticeValue(Value value, LatticeValue &valueEntry,
+  void mergeLatticeValue(FieldRef fieldRef, LatticeValue &valueEntry,
                          LatticeValue source) {
-    if (!source.isOverdefined() && hasDontTouch(value))
+    assert(isRoot(fieldRef) && "value must be known to be root beforehand");
+    if (!source.isOverdefined() && hasDontTouch(fieldRef.getValue()))
```

Comment:
Yeah, it is not field sensitive now so I agree that this is too restrictive.  I will make it only block the field with don't touch.

Reply:
I added a function `bool hasDontTouch(FieldRef fieldRef)` to check the annotation in the field sensitive way. 
But I haven't actually implemented it for now. Because we have to check the donttouch doesn't exist any of parent of given fieldID, but  the current Annotation class doesn't have the utility to check that one fieldID is a parent of another fieldID. It will require moderate amount of code so I'd like to do it in another patch. Therefore, I made it very conservatively, just return whether root has donttouch or not.